diff -Nur gradm-1.1c/Makefile.in gradm-1.1c-new/Makefile.in
--- gradm-1.1c/Makefile.in	Tue Dec  4 01:20:09 2001
+++ gradm-1.1c-new/Makefile.in	Mon Feb  4 16:03:46 2002
@@ -33,11 +33,8 @@
 grhash.o : grhash.c $(DEPS)
 
 install: gradm 
-	$(INSTALL) -m 0700 -o root -g root gradm $(sbindir)
-	$(SHELL) ./mkinstalldirs $(mandir) $(installman) $(grdir)
-	chmod 700 $(grdir)
-	$(sbindir)/gradm -P
-	$(INSTALL) -m 0444 -o root -g root gradm.8 $(installman)
+	$(INSTALL) gradm $(DESTDIR)$(sbindir)
+	$(SHELL) ./mkinstalldirs $(DESTDIR)$(mandir) $(DESTDIR)$(installman) $(DESTDIR)$(grdir)
 
 include/buildstamp.h:
 	echo \#define GR_BUILD_STAMP \"`date`\" > include/buildstamp.h
