ignore "CVS" ".svn" "*~" files/dirs
--- gradm2/gradm.l~	2008-04-04 22:05:25.000000000 +0200
+++ gradm2/gradm.l	2008-07-18 15:06:51.613273610 +0200
@@ -388,7 +388,9 @@
 		return;
 
 	for (i = 0; i < n; i++) {
-		if (!strcmp(namelist[i]->d_name, ".") || !strcmp(namelist[i]->d_name, ".."))
+		if (!strcmp(namelist[i]->d_name, ".") || !strcmp(namelist[i]->d_name, "..") ||
+		    !strcmp(namelist[i]->d_name, ".svn") || !strcmp(namelist[i]->d_name, "CVS") ||
+		    !strcmp(namelist[i]->d_name+strlen(namelist[i]->d_name)-1, "~"))
 			continue;
 		tmp = calloc(1, sizeof(struct include_entry));
 		if (tmp == NULL)
