64bit caps miss in show_policy()
--- gradm2/gradm.l~	2008-07-26 20:21:06.000000000 +0200
+++ gradm2/gradm.l	2008-07-26 20:22:02.208776756 +0200
@@ -553,8 +553,8 @@
 	printf("\n");
 	for (proc = rolp->hash->first;proc;proc=proc->prev) {
 		printf("\tSUBJECT: %s dev:%lu inode:%lu mode:%lu c_raise:%x c_drop:%x\n",
-			proc->filename, proc->dev, proc->inode, proc->mode, ~proc->cap_drop,
-			proc->cap_drop);
+			proc->filename, proc->dev, proc->inode, proc->mode,
+			cap_invert(proc->cap_drop), proc->cap_drop);
 		for (i = 0; i < proc->ip_num; i++) {
 			char ipaddr[4];
 			int c;
