--- gradm-svn/gradm_cap.c	2006-12-17 13:51:48.772510000 +0100
+++ gradm-svn/gradm_cap.c.	2006-12-17 02:52:19.474597750 +0100
@@ -30,6 +30,7 @@
 	{"CAP_LEASE", 28},
 	{"CAP_AUDIT_WRITE", 29},
 	{"CAP_AUDIT_CONTROL", 30},
+	{"CAP_CONTEXT", 31},
 	{"CAP_ALL", ~0}
 };
 
--- gradm-svn/gradm_defs.h~	2006-12-13 01:08:30.356024500 +0100
+++ gradm-svn/gradm_defs.h	2006-12-17 14:00:14.660126000 +0100
@@ -414,7 +414,7 @@ struct gr_arg_wrapper {
 };
 
 extern char *rlim_table[GR_NLIMITS];
-extern struct capability_set capability_list[32];
+extern struct capability_set capability_list[33];
 extern struct paxflag_set paxflag_list[5];
 
 extern int is_24_kernel;
